### PR TITLE
Allow Whitelisting of Sysctls Configuration

### PR DESCRIPTION
**Pull request by mburke5678**
https://jira.coreos.com/browse/OSDOCS-502
https://jira.coreos.com/browse/PROD-1038
https://jira.coreos.com/browse/NODE-140

### CODE REVIEW

File: modules/nodes-containers-sysctls-about.adoc [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+Additionally, most of the sysctls in the *_net.*_* group are known
+to be namespaced. Their namespace adoption differs based on the kernel
+version and distributor.
+
+To check which *_net.*_* sysctls are namespaced on your system, run the
+following command:
+
+----
+$ podman run --rm -ti docker.io/fedora \
```
Comment (rphillips) [inline modules/nodes-containers-sysctls-about.adoc]: this section on running podman is awesome, but is somewhat contrary to us not wanting customers to SSH into machines. I vote on removing this blurb on checking the namespaced sysctls.

File: modules/nodes-containers-sysctls-unsafe.adoc [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+----
+$ oc edit machineconfigpool worker
+
+apiVersion: machineconfiguration.openshift.io/v1
+kind: MachineConfigPool
+metadata:
+  creationTimestamp: 2019-02-08T14:52:39Z
+  generation: 1
+  labels:
+    custom-kubelet: small-pods <1>
```
Comment (rphillips) [inline modules/nodes-containers-sysctls-unsafe.adoc]: The label here should be `custom-kubelet: sysctl`

**Review by rphillips — CHANGES_REQUESTED** [review]
Looks great overall... Just a couple comments.

**Comment by mburke5678** [comment]
@rphillips I added a section to `nodes-containers-sysctls-unsafe.adoc` with steps to verify that the cluster is rebooted and ready before creating the pod with unsafe sysctls. Can you verify if these are the best ways to verify the reboot?
@vikram-redhat I am not sure who the br QE resource is. Could I bother you to assign this? 

I am on PTO until 8/16 and won't be checking email until 8/15 at earliest. I would like to get the procedures in this PR reviewd. Please don't worry about obvious typos. I will fix them when I get back.

**Comment by xltian** [comment]
@sunilcio from QE will take a look at this.

**Review by rphillips — APPROVED** [review]

File: modules/nodes-containers-sysctls-unsafe.adoc [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+Due to their nature of being unsafe, the use of unsafe sysctls is
+at-your-own-risk and can lead to severe problems like wrong behavior of
+containers, resource shortage, or complete breakage of a node.
+====
+
+.Procedure
+
+. Add a label to the Machine Config Pool where the containers where containers with the unsafe sysctls will run:
++
+source,yaml]
```
Comment (sunilcio) [inline modules/nodes-containers-sysctls-unsafe.adoc]: Small syntax error. `[source,yaml]`

**Comment by mburke5678** [comment]
@openshift/team-documentation PTAL.
I am replacing the _sysctl_ topic from 3.11 back into 4.2 and modularizing.  It appears that the original topic was pulled from upstream in large part. I tried cleaning things up a bit. But i don't know sysctl. 
I likely could have found more introductory info. Not sure it's needed though, as I am assuming that any user who would use sysctl would understand what they are doing.

File: modules/nodes-containers-sysctls-about.adoc [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+requirement for sysctls to be accessible in a pod context within Kubernetes.
+
+The following sysctls are known to be namespaced:
+
+- *_kernel.shm*_*
+- *_kernel.msg*_*
+- *_kernel.sem_*
+- *_fs.mqueue.*_*
+
+Additionally, most of the sysctls in the **net.*** group are known
```
Comment (adellape) [inline modules/nodes-containers-sysctls-about.adoc]: Check markup choice here.

File: modules/nodes-containers-sysctls-about.adoc [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+Sysctls are grouped into _safe_ and _unsafe_ sysctls. 
+
+For a sysctl to be considered safe, it must use proper
+namespacing and must be properly isolated between pods on the same
+node. This means that if you set a sysctl for one pod it must not:
+
+- Influence any other pod on the node
+- Harm the node's health
+- Gain CPU or memory resources outside of the resource limits of a pod
```
Comment (adellape) [inline modules/nodes-containers-sysctls-about.adoc]: Periods at the end of these?

File: modules/nodes-containers-sysctls-setting.adoc [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+// * nodes/containers/nodes-containers-sysctls.adoc
+
+[id="nodes-containers-sysctls-setting_{context}"]
+= Setting sysctls for a pod
+
+You can set sysctls on pods using the pod's `securityContext`. The `securityContext`
+applies to all containers in the same pod.
+
+Safe sysctls are allowed by default. A pod with unsafe sysctls will fail 
+to launch on any node unless the admin explicitly enables unsafe sysctls for
```
Comment (adellape) [inline modules/nodes-containers-sysctls-setting.adoc]: s/will fail/fails/
s/admin/cluster administrator/

File: modules/nodes-containers-sysctls-unsafe.adoc [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+[WARNING]
+====
+Due to their nature of being unsafe, the use of unsafe sysctls is
+at-your-own-risk and can lead to severe problems, such as improper behavior of
+containers, resource shortage, or breaking a node.
+====
+
+.Procedure
+
+. Add a label to the Machine Config Pool where the containers where containers
```
Comment (adellape) [inline modules/nodes-containers-sysctls-unsafe.adoc]: MachineConfigPool?

File: modules/nodes-containers-sysctls-unsafe.adoc [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+spec:
+  machineConfigPoolSelector:
+    matchLabels:
+      custom-kubelet: sysctl <1>
+  kubeletConfig:
+    allowedUnsafeSysctls: <2>
+      - "kernel.msg*"
+      - "net.ipv4.route.min_pmtu"
+----
+<1> Specify the label from the Machine Config Pool.
```
Comment (adellape) [inline modules/nodes-containers-sysctls-unsafe.adoc]: MachineConfigPool?

File: modules/nodes-containers-sysctls-unsafe.adoc [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+at-your-own-risk and can lead to severe problems, such as improper behavior of
+containers, resource shortage, or breaking a node.
+====
+
+.Procedure
+
+. Add a label to the Machine Config Pool where the containers where containers 
+with the unsafe sysctls will run:
++
+[source,yaml]
```
Comment (adellape) [inline modules/nodes-containers-sysctls-unsafe.adoc]: Maybe don't add this syntax highlighting if the other YAML blocks aren't going to get it, for consistency within the assembly at least. With the combination of the CLI command + YAML output, it makes the command look wacky, too.

File: modules/nodes-containers-sysctls-unsafe.adoc [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+        rendered-worker-ccbfb5d2838d65013ab36300b7b3dc13
+      reason: ''
+      status: 'True'
+      type: Updated
+----
+
+. When the cluster is ready, check for the merged `KubeletConfig` in the new MachineConfig:
++
+----
+oc get machineconfig 99-worker-XXXXXX-XXXXX-XXXX-XXXXX-kubelet -o json | grep ownerReference -A7
```
Comment (adellape) [inline modules/nodes-containers-sysctls-unsafe.adoc]: `$` prompt at front, and maybe add a blank line after the command to give it some room from the output.

**Review by adellape — COMMENTED** [review]
Minor suggestions. :+1:

**Comment by mburke5678** [comment]
@adellape Thanks for the review!!

**Pull request merged**